### ISSUE TITLE
Typescript error with required properties

### ISSUE THREAD

**Issue by stevekrouse** [nvie/decoders#750]
It seems like `AllowImplicit` is confusing Typescript into thinking that required properties are optional.

```ts
interface A {
  a: string;
}

const aDecoder: Decoder<A> = object({
  a: string
});
```

Output:

```
Type 'Decoder<{ a?: string; }, unknown>' is not assignable to type 'Decoder<A, unknown>'.
  Type '{ a?: string; }' is not assignable to type 'A'.
    Property 'a' is optional in type '{ a?: string; }' but required in type 'A'.ts(2322)
```

Live example: https://codesandbox.io/s/nostalgic-sinoussi-lkrq1?file=/src/index.ts

**Comment by nvie** [comment]
Thank you for this bug report @stevekrouse and thanks for these amazing reproduction steps @djlauk 🙏 !

I've narrowed the source of the bug down further and found that the `strictNullChecks` flag is the real culprit (which is part of the family that `strict` will enable).

https://www.typescriptlang.org/tsconfig#strictNullChecks

> _"When `strictNullChecks` is false, `null` and `undefined` are effectively ignored by the language. This can lead to unexpected errors at runtime."_

This sounds really scary and I'm looking to understand when you would ever want to use this, especially if you are already in the center of the Venn diagram of users interested in using TypeScript and decoders 😄 

I'm currently trying to figure out if I want to dive in and see what's causing this and see if I can somehow make it compatible with that setting (it would be nice it it would Just Work™ no matter how your TypeScript project is configured of course!), or… if I should just recommend to always use `strictNullChecks: true` when using decoders.

Thoughts? Are you deliberately not using `strictNullChecks` for some reason?

**Comment by djlauk** [comment]
@nvie Thanks for looking into the matter!

To me, it would be sufficient to add it to the documentation, I guess. I can try and come up with a PR for that.

I encountered this problem when adding decoders to a toy side project whilst trying out [SvelteKit](https://kit.svelte.dev). Turns out their default `tsconfig.json` doesn't enable strict mode.

My best guess about people not wanting TS's strict mode, is to ease migration from JS. In the case of Svelte / SvelteKit the destructuring shorthand is conveniently used for function parameters. In JS that looks nicely terse (e.g. have a look at the `get` function [in this example here](https://kit.svelte.dev/docs#routing-endpoints)); in TS that terseness only works without strict mode. (If you enable strict mode, TS complains about `params` implicitly being type `any`.)
